src/generator.rs

```diff
@@ -0,0 +1,201 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Wake, Waker};
+
+use url::Url;
+
+use crate::network_client::NetworkProtocol;
+use crate::{Error, InitializeSecurityContextResult};
+
+pub struct Interrupt<YieldTy, ResumeTy> {
+    value_to_yield: Option<YieldTy>,
+    yielded_value: YieldedValue<YieldTy>,
+    resumed_value: ResumedValue<ResumeTy>,
+    ready_to_resume: bool,
+}
+
+impl<'a, YieldTy, ResumeTy> Future for Interrupt<YieldTy, ResumeTy>
+where
+    YieldTy: Unpin,
+{
+    type Output = ResumeTy;
+
+    fn poll(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        if this.ready_to_resume {
+            let resumed_value = this.resumed_value.try_lock().unwrap().take().unwrap();
+            Poll::Ready(resumed_value)
+        } else {
+            let value_to_yield = this.value_to_yield.take().unwrap();
+            *this.yielded_value.try_lock().unwrap() = Some(value_to_yield);
+            this.ready_to_resume = true;
+            Poll::Pending
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct YieldPoint<YieldTy, ResumeTy> {
+    yielded_value: YieldedValue<YieldTy>,
+    resumed_value: ResumedValue<ResumeTy>,
+}
+
+impl<'a, YieldTy, ResumeTy> Clone for YieldPoint<YieldTy, ResumeTy> {
+    fn clone(&self) -> Self {
+        Self {
+            yielded_value: self.yielded_value.clone(),
+            resumed_value: self.resumed_value.clone(),
+        }
+    }
+}
+
+impl<'point, YieldTy, ResumeTy> YieldPoint<YieldTy, ResumeTy> {
+    pub fn suspend(&'point self, value: YieldTy) -> Interrupt<YieldTy, ResumeTy> {
+        Interrupt {
+            value_to_yield: Some(value),
+            yielded_value: Arc::clone(&self.yielded_value),
+            resumed_value: Arc::clone(&self.resumed_value),
+            ready_to_resume: false,
+        }
+    }
+}
+
+type YieldedValue<T> = Arc<Mutex<Option<T>>>;
+type ResumedValue<T> = Arc<Mutex<Option<T>>>;
+type PinnedFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;
```

Comment:
Is there anything preventing this to be `Sync` in our case?
```suggestion
type PinnedFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + Sync + 'a>>;
```

Reply:
not really, removed